core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java

```diff
@@ -118,53 +129,55 @@ void executeSearch(Queue<SearchRequestSlot> requests, AtomicArray<MultiSearchRes
         }
 
         /*
-         * With a request in hand, we are going to asynchronously execute the search request. When the search request returns, either with
-         * a success or with a failure, we set the response corresponding to the request. Then, we enter a loop that repeatedly pulls
-         * requests off the request queue, this time only setting the response corresponding to the request.
+         * With a request in hand, we are now prepared to execute the search request. There are two possibilities, either we go asynchronous
+         * or we do not (this can happen if the request does not resolve to any shards). If we do not go asynchronous, we are going to come
+         * back on the same thread that attempted to execute the search request. At this point, or any other point where we come back on the
+         * same thread as when the request was submitted, we should not recurse lest we might descend into a stack overflow. To avoid this,
+         * when we handle the response rather than going recursive, we set ourselves up to loop around and submit a new request. Otherwise
+         * we recurse.
          */
-        searchAction.execute(request.request, new ActionListener<SearchResponse>() {
-            @Override
-            public void onResponse(final SearchResponse searchResponse) {
-                handleResponse(request.responseSlot, new MultiSearchResponse.Item(searchResponse, null));
-                executeSearchLoop();
-            }
-
-            @Override
-            public void onFailure(final Exception e) {
-                handleResponse(request.responseSlot, new MultiSearchResponse.Item(null, e));
-                executeSearchLoop();
+        final AtomicReference<SearchRequestSlot> next = new AtomicReference<>();
+        final Thread thread = Thread.currentThread();
+        do {
+            final SearchRequestSlot current;
+            if (next.get() != null) {
```

Comment:
mayb initialize next to the initial request? I think this will mean that we can remove this if and make do with `current = next.getAndSet(null)`